File: colorworld-ui/src/app/demo/domain/formula.ts
export class FormulaHeader {
    brand: string;
    model: string;
    shade: string;
    variant: string;
    code: string;
}

export class FormulaBody {
    code: string;
    ingredients: string;
    weight: number;
    price: number;
}

export class Formula extends FormulaHeader{
    formula: FormulaBody[];
}

File: src/main/java/com/mrajupaint/colorworld/formula/entity/Formula.java
package com.mrajupaint.colorworld.formula.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.IdClass;
import jakarta.persistence.Table;
import lombok.Data;

@Table(name = "formula")
@Entity
@IdClass(FormulaKey.class)
@Data
public class Formula {

	@Id
	@Column
	private String code;
	
	@Id
	@Column
	private String ingredients;
	
	@Column
	private double weight;
	
	@Column
	private double price;
}
